Userland/Libraries/LibWeb/CSS/StyleValues/CalculatedStyleValue.cpp

```diff
@@ -390,6 +390,71 @@ ErrorOr<void> InvertCalculationNode::dump(StringBuilder& builder, int indent) co
     return {};
 }
 
+ErrorOr<NonnullOwnPtr<ConstantCalculationNode>> ConstantCalculationNode::create(ConstantType constant)
+{
+    return adopt_nonnull_own_or_enomem(new (nothrow) ConstantCalculationNode(constant));
+}
+
+ConstantCalculationNode::ConstantCalculationNode(ConstantType constant)
+    : CalculationNode(Type::Constant)
+    , m_constant(constant)
+{
+}
+
+ConstantCalculationNode::~ConstantCalculationNode() = default;
+
+ErrorOr<String> ConstantCalculationNode::to_string() const
+{
+    switch (m_constant) {
+    case CalculationNode::ConstantType::E:
+        return "e"_short_string;
+    case CalculationNode::ConstantType::PI:
+        return "pi"_short_string;
+    case CalculationNode::ConstantType::Infinity:
+        return String::from_utf8("infinity"sv);
+    case CalculationNode::ConstantType::MinusInfinity:
+        return String::from_utf8("-infinity"sv);
+    case CalculationNode::ConstantType::NaN:
+        return String::from_utf8("NaN"sv);
+    }
+
+    VERIFY_NOT_REACHED();
+}
+
+Optional<CalculatedStyleValue::ResolvedType> ConstantCalculationNode::resolved_type() const
+{
+    return CalculatedStyleValue::ResolvedType::Number;
+}
+
+CalculatedStyleValue::CalculationResult ConstantCalculationNode::resolve([[maybe_unused]] Layout::Node const* layout_node, [[maybe_unused]] CalculatedStyleValue::PercentageBasis const& percentage_basis) const
+{
+    switch (m_constant) {
+    case CalculationNode::ConstantType::E:
+        return { Number(Number::Type::Number, M_E) };
+    case CalculationNode::ConstantType::PI:
+        return { Number(Number::Type::Number, M_PI) };
+    case CalculationNode::ConstantType::Infinity:
+        return { Number(Number::Type::Number, NumericLimits<float>::max()) };
+    case CalculationNode::ConstantType::MinusInfinity:
+        return { Number(Number::Type::Number, NumericLimits<float>::lowest()) };
```

Comment:
technically `max` and `lowest` do not denote the representation of infinities… or do they? 🤔

(should those be `INFINITY` and `-INFINITY` resp.?)

Reply:
This is tricky... `infinity - 7` should equal `infinity`, not `FLOAT_MAX - 7`. And similarly, `infinity - infinity` should be `NaN`. So using the actual infinity constants feels correct. But also, from what I can tell we *should* clamp the end result of a calculation tree to the range. (And Chrome seems to agree.)

Reply:
It feels like we need to track `NaN` and `Infinity` across every node do resolve it correctly, but with this PR i only wanted to parse the constants, we can improve on handling them correctly later

Reply:
Sure. If you can add a FIXME there though that would be good!